templates/cancel.php

```diff
@@ -1,35 +1,119 @@
 <?php
 /**
+ * Template: Cancel
+ *
  * Overrides ../wp-content/plugins/paid-memberships-pro/templates/cancel.php
+ *
+ * @version 2.0
+ *
+ * @author Paid Memberships Pro
  */
-global $pmpro_msg, $pmpro_msgt, $pmpro_confirm;
+global $pmpro_msg, $pmpro_msgt, $pmpro_confirm, $current_user, $wpdb;
 
-if ( $pmpro_msg ) {
-?>
-<div class="pmpro_message <?php echo esc_html( $pmpro_msgt ); ?>"><?php echo esc_html( $pmpro_msg ); ?></div>
-<?php
-}
-?>
+if ( isset( $_REQUEST['levelstocancel'] ) && $_REQUEST['levelstocancel'] !== 'all' ) {
+	//convert spaces back to +
+	$_REQUEST['levelstocancel'] = str_replace( array( ' ', '%20' ), '+', $_REQUEST['levelstocancel'] );
 
-<?php if ( ! $pmpro_confirm ) { ?>           
+	//get the ids
+	$old_level_ids = array_map( 'intval', explode( "+", preg_replace( "/[^0-9al\+]/", "", $_REQUEST['levelstocancel'] ) ) );
 
-<form action="" method="post">
-
-<p><?php esc_html_e( 'Are you sure you want to cancel your membership?', 'pmpro-reason-for-cancelling' ); ?></p>
-<p>
-<?php esc_html_e( 'If so, please enter a reason for cancelling and click "Yes, cancel my account" below.', 'pmpro-reason-for-cancelling' ); ?><br />
-<textarea name="reason"></textarea>
-</p>
-
-<p>	
-	<input type="hidden" name="membership_cancel" value="1" />
-	<input type="hidden" name="confirm" value="1" />
-	<input type="submit" name="submit" value="<?php esc_html_e( 'Yes, cancel my account', 'pmpro-reason-for-cancelling' ); ?>" />
-	|
-	<a class="pmpro_nolink nolink" href="<?php echo pmpro_url( 'account' ); ?>"><?php esc_html_e( 'No, keep my account', 'pmpro-reason-for-cancelling' ); ?></a>
-</p>
-<?php } else { ?>
-	<p><a href="<?php echo get_home_url(); ?>"><?php esc_html_e( 'Click here to go to the home page.', 'pmpro-reason-for-cancelling' ); ?></a></p>
-<?php } ?>
+} elseif ( isset( $_REQUEST['levelstocancel'] ) && $_REQUEST['levelstocancel'] == 'all') {
+	$old_level_ids = 'all';
+} else {
+	$old_level_ids = false;
+}
+?>
+<div id="pmpro_cancel" class="<?php echo pmpro_get_element_class( 'pmpro_cancel_wrap', 'pmpro_cancel' ); ?>">
+	<?php
+		if ( $pmpro_msg ) {
+			?>
+			<div class="<?php echo pmpro_get_element_class( 'pmpro_message ' . $pmpro_msgt, $pmpro_msgt ); ?>"><?php echo $pmpro_msg?></div>
+			<?php
+		}
+	?>
+	<?php
+		if ( ! $pmpro_confirm ) {
+			if ( $old_level_ids ) {
+				?>
+				<form action="" method="post">
+				<?php
+				if ( ! is_array( $old_level_ids ) && $old_level_ids == "all" ) {
+					?>
+					<p><?php _e( 'Are you sure you want to cancel your membership?', 'pmpro-reason-for-cancelling' ); ?></p>
+					<?php
+				} else {
+					$level_names = $wpdb->get_col( "SELECT name FROM $wpdb->pmpro_membership_levels WHERE id IN('" . implode( "','", $old_level_ids ) . "')" );
+					?>
+					<p><?php printf( _n( 'Are you sure you want to cancel your %s membership?', 'Are you sure you want to cancel your %s memberships?', count( $level_names ), 'pmpro-reason-for-cancelling' ), pmpro_implodeToEnglish( $level_names ) ); ?></p>
```

Comment:
Needs `// translators: %s: The level name(s).` added to php line above the _n call here